core/languages.lua

```diff
@@ -9,30 +13,124 @@ end)
+        -- It matched "en" eventually (as we don't have yet an "languages.en-GB" resources)
+        -- PROBLEM: Our languages.xxx files (almost) all work by side effects, putting various things,
+        -- in the case of our example, in SILE.nodeMarkers.en, SILE.hyphenator.languages.en
+        -- and SU.formatNumber.en... While we now expect the language to be "en-GB"...
+        -- It's a HACK, but copy the stuff into our language.
+        SILE.nodeMakers[language] = SILE.nodeMakers[matchedlang]
+        SU.formatNumber[language] = SU.formatNumber[matchedlang]
+        SILE.hyphenator.languages[language] = SILE.hyphenator.languages[matchedlang]
+      end
     end
-    local ftlresource = string.format("i18n.%s", language)
-    SU.debug("fluent", "Loading FTL resource", ftlresource, "into locale", language)
-    fluent:set_locale(language)
-    local gotftl, ftl = pcall(require, ftlresource)
-    if not gotftl then
-      SU.warn(("Unable to load localized strings (e.g. table of contents header text) for %s: %s")
-        :format(language, ftl:gsub(":.*", "")))
+
+    -- We need to find fluent reources for this BCP47 identifier, from the less specific
+    -- to the more general.
+    local ftlresource, matchedi18n = forLanguage(language, function (lang)
+      local resource = string.format("i18n.%s", lang)
+      SU.debug("fluent", "Loading FTL resource", resource, "into locale", lang)
+      fluent:set_locale(lang)
+      local gotftl, ftl = pcall(require, resource)
+      return gotftl and ftl
+    end)
+    if not ftlresource then
+      SU.warn(("Unable to load localized strings (e.g. table of contents header text) for %s")
+        :format(language))
+    else
+      print(("Load localized strings for %s: matched %s") -- HACK We'll need a mere SU.debug when OK...
+        :format(language, matchedi18n))
+      if language ~= matchedi18n then
+        -- Now that's even more UGLY. Say the input language was "en-GB".
+        -- It matched "en" eventually (as we don't have yet an "i18n.en-GB" resources)
+        -- PROBLEM: the fluent locale must be set to the target language before loading
+        -- a ftl file. APIs that aren't stateless are messy :(
+        -- in the case of our example, they had to be read into "en"...
+        -- HACK HACK, all we can do is reloaad it fully, but under the target "en-GB" name...
+        local loaded = string.format("i18n.%s", matchedi18n)
+        package.loaded[loaded] = nil -- HACK force reload!!!
+        fluent:set_locale(language)
+        require(string.format("i18n.%s", matchedi18n))
+      end
     end
-    if type(lang) == "table" and lang.init then
-      lang.init()
+
+    -- Most language resource files act by side effects, directly tweaking
+    -- SILE.nodeMarkers.xx, SILE.hyphenator.languages.xx, SU.formatNumber.xx (etc.? Doh)
+    -- BUT some don't do that exactly AND return a table with an init method...
+    -- Unclear API discrepancy, heh.
```

Comment:
Yes, the ones that act by side effect predate my attempts to bring some modularity to the interfaces. We need to switch to a module system that can subclass other modules when desired. We've done it for document classes, packages, inputters, outputters, typesetters and more: but not yet languages. We need to.